File: zenoh/src/main/java/org/eclipse/zenoh/net/Subscriber.java
/*
 * Copyright (c) 2017, 2020 ADLINK Technology Inc.
 *
 * This program and the accompanying materials are made available under the
 * terms of the Eclipse Public License 2.0 which is available at
 * http://www.eclipse.org/legal/epl-2.0, or the Apache License, Version 2.0
 * which is available at https://www.apache.org/licenses/LICENSE-2.0.
 *
 * SPDX-License-Identifier: EPL-2.0 OR Apache-2.0
 *
 * Contributors:
 *   ADLINK zenoh team, <zenoh@adlink-labs.tech>
 */
package org.eclipse.zenoh.net;

import org.eclipse.zenoh.core.ZException;
import org.eclipse.zenoh.swig.zenohc;
import org.eclipse.zenoh.swig.zn_sub_t;

/**
 * A Subscriber (see
 * {@link Session#declareSubscriber(String, SubMode, SubscriberCallback)}).
 */
public class Subscriber {

    private zn_sub_t sub;

    protected Subscriber(zn_sub_t sub) {
        this.sub = sub;
    }

    /**
     * Pull data for the {@link SubMode.Kind#ZN_PULL_MODE} or
     * {@link SubMode.Kind#ZN_PERIODIC_PULL_MODE} subscribtion. The pulled data will
     * be provided by calling the
     * {@link DataHandler#handleData(String, java.nio.ByteBuffer, DataInfo)}
     * function provided to the
     * {@link Session#declareSubscriber(String, SubMode, DataHandler)} function.
     * 
     * @throws ZException if pull failed.
     */
    public void pull() throws ZException {
        int result = zenohc.zn_pull(sub);
        if (result != 0) {
            throw new ZException("zn_pull failed", result);
        }
    }

    /**
     * Undeclare the Subscriber.
     * 
     * @throws ZException if undeclaration failed.
     */
    public void undeclare() throws ZException {
        int error = zenohc.zn_undeclare_subscriber(sub);
        if (error != 0) {
            throw new ZException("zn_undeclare_subscriber failed ", error);
        }
    }

}

File: zenoh-java/src/commonMain/kotlin/io/zenoh/pubsub/Subscriber.kt
//
// Copyright (c) 2023 ZettaScale Technology
//
// This program and the accompanying materials are made available under the
// terms of the Eclipse Public License 2.0 which is available at
// http://www.eclipse.org/legal/epl-2.0, or the Apache License, Version 2.0
// which is available at https://www.apache.org/licenses/LICENSE-2.0.
//
// SPDX-License-Identifier: EPL-2.0 OR Apache-2.0
//
// Contributors:
//   ZettaScale Zenoh Team, <zenoh@zettascale.tech>
//

package io.zenoh.pubsub

import io.zenoh.handlers.BlockingQueueHandler
import io.zenoh.jni.JNISubscriber
import io.zenoh.keyexpr.KeyExpr
import io.zenoh.session.SessionDeclaration

/**
 * A subscriber that allows listening to updates on a key expression and reacting to changes.
 *
 * Its main purpose is to keep the subscription active as long as it exists.
 *
 * Example using the default [BlockingQueueHandler] handler:
 *
 * ```java
 * var queue = session.declareSubscriber("a/b/c");
 * try (Session session = Zenoh.open(config)) {
 *     try (var subscriber = session.declareSubscriber(keyExpr)) {
 *         var receiver = subscriber.getReceiver();
 *         assert receiver != null;
 *         while (true) {
 *             Optional<Sample> wrapper = receiver.take();
 *             if (wrapper.isEmpty()) {
 *                 break;
 *             }
 *             System.out.println(wrapper.get());
 *         }
 *     }
 * }
 * ```
 *
 * Example using a callback:
 * ```java
 * try (Session session = Zenoh.open(config)) {
 *     session.declareSubscriber(keyExpr, System.out::println);
 * }
 * ```
 *
 * Example using a handler:
 * ```java
 * class MyHandler implements Handler<Sample, ArrayList<Sample>> {...}
 *
 * //...
 * try (Session session = Zenoh.open(config)) {
 *     var handler = new MyHandler();
 *     var arraylist = session.declareSubscriber(keyExpr, handler);
 *     // ...
 * }
 * ```
 */
sealed class Subscriber(
    val keyExpr: KeyExpr, private var jniSubscriber: JNISubscriber?
) : AutoCloseable, SessionDeclaration {

    fun isValid(): Boolean {
        return jniSubscriber != null
    }

    override fun undeclare() {
        jniSubscriber?.close()
        jniSubscriber = null
    }

    override fun close() {
        undeclare()
    }

    protected fun finalize() {
        jniSubscriber?.close()
    }
}

/**
 * Subscriber using a callback to handle incoming samples.
 *
 * Example:
 * ```java
 * try (Session session = Zenoh.open(config)) {
 *     session.declareSubscriber(keyExpr, System.out::println);
 * }
 * ```
 */
class CallbackSubscriber internal constructor(keyExpr: KeyExpr, jniSubscriber: JNISubscriber?): Subscriber(keyExpr, jniSubscriber)

/**
 * Subscriber using a [io.zenoh.handlers.Handler] for handling incoming samples.
 *
 * Example using the default handler:
 * ```java
 * try (Session session = Zenoh.open(config)) {
 *     try (HandlerSubscriber<BlockingQueue<Optional<Sample>>> subscriber = session.declareSubscriber(keyExpr)) {
 *         BlockingQueue<Optional<Sample>> receiver = subscriber.getReceiver();
 *         assert receiver != null;
 *         while (true) {
 *             Optional<Sample> wrapper = receiver.take();
 *             if (wrapper.isEmpty()) {
 *                 break;
 *             }
 *             System.out.println(wrapper.get());
 *         }
 *     }
 * }
 * ```
 *
 * @param R The type of the receiver.
 * @param receiver The receiver of the subscriber's handler.
 */
class HandlerSubscriber<R> internal constructor(keyExpr: KeyExpr, jniSubscriber: JNISubscriber?, val receiver: R): Subscriber(keyExpr, jniSubscriber)
